test/unit/org/apache/cassandra/index/sai/cql/NumericIndexMixedVersionTest.java

```diff
@@ -0,0 +1,146 @@
+            // Query that will hit all sstables and exceed the cassandra.sai.partition_row_batch_size limit
+            var rows = executeNetWithPaging("SELECT pk FROM %s WHERE val >= 0 LIMIT 10000", batchLimit / 2);
+            assertEquals(expectedRows, rows.all().size());
+
+            rows = executeNetWithPaging("SELECT pk FROM %s WHERE val >= 0 LIMIT 10000", batchLimit);
+            assertEquals(expectedRows, rows.all().size());
+
+            rows = executeNetWithPaging("SELECT pk FROM %s WHERE val >= 0 LIMIT 10000", batchLimit * 2);
+            assertEquals(expectedRows, rows.all().size());
+
+            // Test without paging
+            assertNumRows(expectedRows, "SELECT pk FROM %%s WHERE val >= 0 LIMIT 10000");
+        });
+    }
+
+    @Test
+    public void testMultiVersionCompatibilityWithClusteringColumns() throws Throwable
+    {
+        createTable("CREATE TABLE %s (pk int, ck int, val int, PRIMARY KEY(pk, ck)) WITH CLUSTERING ORDER BY (ck ASC)");
+        createIndex("CREATE CUSTOM INDEX ON %s(val) USING 'StorageAttachedIndex'");
+
+        // Note that we do not test the multi-version path where compaction produces different sstables, which is
+        // the norm in CNDB. If we had a way to compact individual sstables, we could.
+        disableCompaction();
+
+        SAIUtil.setCurrentVersion(Version.AA);
+        int ck = 0;
+        for (int j = 0; j < 500; j++)
+            execute("INSERT INTO %s (pk, ck, val) VALUES (1, ?, ?)", ck++, j);
+        flush();
+        compact();
+
+        // Insert 500 rows per version
+        for (var version : VERSIONS)
+        {
+            SAIUtil.setCurrentVersion(version);
+            for (int j = 0; j < 500; j++)
+                execute("INSERT INTO %s (pk, ck, val) VALUES (1, ?, ?)", ck++, j);
+            flush();
+        }
+
+        // Confirm that compaction (aka rebuilding all indexes onto same version) also produces correct results
+        final int expectedRows = ck;
+        runThenFlushThenCompact(() -> {
+            // When using paging, we get an excessive number of results because of logic within the contoller.select
+            // method that short circuits when one of the indexes is aa (not row aware).
+            var batchLimit = CassandraRelevantProperties.SAI_PARTITION_ROW_BATCH_SIZE.getInt();
+            var rows = executeNetWithPaging("SELECT ck FROM %s WHERE val >= 0 LIMIT 10000", batchLimit / 2);
+            assertEquals(expectedRows, rows.all().size());
+
+            rows = executeNetWithPaging("SELECT ck FROM %s WHERE val >= 0 LIMIT 10000", batchLimit);
+            assertEquals(expectedRows, rows.all().size());
+
+            rows = executeNetWithPaging("SELECT ck FROM %s WHERE val >= 0 LIMIT 10000", batchLimit * 2);
+            assertEquals(expectedRows, rows.all().size());
+
+            // Test without paging. This test actually fails by producing fewer than expected rows because of an issue in
```

Comment:
> This test actually fails

I cannot see the fail in CI

Reply:
It fails before the fix. I guess the formulation is confusing.